src/vehicle.cpp

```diff
@@ -2787,27 +2784,33 @@ std::vector<int> vehicle::boarded_parts() const
     std::vector<int> res;
     for (size_t p = 0; p < parts.size(); p++) {
         if (part_flag (p, VPFLAG_BOARDABLE) &&
-                parts[p].has_flag(vehicle_part::passenger_flag)) {
+            get_passenger (p)) {
             res.push_back ((int)p);
         }
     }
     return res;
 }
 
-player *vehicle::get_passenger(int p) const
+template<typename T>
+T *vehicle::get_passenger(int p) const
 {
     p = part_with_feature (p, VPFLAG_BOARDABLE, false);
-    if (p >= 0 && parts[p].has_flag(vehicle_part::passenger_flag))
+    if (p >= 0)
     {
-     const int player_id = parts[p].passenger_id;
-     if( player_id == g->u.getID()) {
-      return &g->u;
-     }
-        return g->npc_by_id( player_id );
+	T *psg = g->critter_at<T> (global_part_pos3(p));
```

Comment:
This is relatively costly and will get called a lot. If you want it be done this way, make sure it works reasonably fast in rarer edge cases:
* 20+ cars in reality bubble
* A main battle tank (from the tanks mod) sized vehicle
* 3+ NPCs boarding the above vehicle
* 50+ zombies in reality bubble
* All of the above at once

The problem is, your approach can't be easily optimized. Any slowdown here will surely stay with us.